datafusion/physical-expr/src/aggregate/min_max.rs

```diff
@@ -541,22 +543,38 @@ pub fn max_row(index: usize, accessor: &mut RowAccessor, s: &ScalarValue) -> Res
 #[derive(Debug)]
 pub struct MaxAccumulator {
     max: ScalarValue,
+    moving_max: moving_min_max::MovingMax<ScalarValue>,
 }
 
 impl MaxAccumulator {
     /// new max accumulator
     pub fn try_new(datatype: &DataType) -> Result<Self> {
         Ok(Self {
             max: ScalarValue::try_from(datatype)?,
+            moving_max: moving_min_max::MovingMax::<ScalarValue>::new(),
         })
     }
 }
 
 impl Accumulator for MaxAccumulator {
     fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
-        let values = &values[0];
-        let delta = &max_batch(values)?;
-        self.max = max(&self.max, delta)?;
+        for idx in 0..values[0].len() {
```

Comment:
I am concerned about adding this code to `MinAccumulator` and `MaxAccumulator` as it is on the performance critical path for aggregating in general. 

Specifically, I worry it would slow down queries like this 

```sql
select MIN(time) from my_table_with_10000000_rows;
```

New work that is done
1. A ScalarValue is created for *each* input element (not just the maximum value)
2. The `moving_max` structure is updated using for each input element

Unless we can show performance benchmarks that this additional overhead doesn't matter I don't think we should merge this particular change.

Instead, I think we need to special case here and provide Min/Max implementations (as you have here) such as `RetractableMinAccumulator` and `RetractableMaxAccumulator`  that work for bounded windows and not change the existing Min/Max implementations that are fast for unbounded windows

One way might be in the physical planner when the window bounds are known:

https://github.com/apache/arrow-datafusion/blob/fddb3d3651041f41d66a801f10e27387e84374f7/datafusion/core/src/physical_plan/windows/mod.rs#L57-L76

Another way might be to make a new built in `AggregrateFunction` for min/max that is only used for window functions

Reply:
Thanks for the review @alamb, as in our previous disccussion when @Ted-Jiang proposed an algorithm. We have updated our design to address your concern. For expressions that do not require sliding window implementations(such your example) we use `ForwardAggregateWindowExpr`. This implementation uses more optimized implementation (since it doesn't assume retract). Currently optimized algorithms are supported for accumulators having `row_accumulator_supported` boolean `true`.

Reply:
@mustafasrepo  are you saying that you believe this PR will not negatively impact the performance of a query such as the following?


```sql
select MIN(time) from my_table_with_10000000_rows;
```

Reply:
I missed this point sorry, while answering I had in mind a query such as `MIN(c12) OVER (ORDER BY C12 RANGE BETWEEN UNBOUNDED PRECEDING AND 0.2 FOLLOWING`. This query was running with optimized implementation, however, your example was running with sliding implementation(with overhead). I sent a commit to fix this issue. Now your example should be running a non-sliding implementation. Thanks for pointing it out. I have followed your suggestion to solve this issue (Name of the new accumulators are `SlidingMinAccumulator` and `SlidingMaxAccumulator`).